File: ui/src/core/audit.ts
// The audit response: GET /api/executions/{id} (core/server's Audit, wrapping
// core/replay.Timeline). This is what the live event stream (core/live.ts)
// cannot carry — the frozen Workflow and its Workers (so a node's Goal and
// Contract render, REQ-UI-03), plus each node's actual artifact bytes
// (REQ-UI-04) — not just the hash/type ArtifactCreated puts on the wire.
//
// Two views of the same execution coexist deliberately (PRIN-02): `live.ts`
// folds the event stream into evolving per-node status for the Gantt/canvas;
// this module is the frozen-at-start-plus-final-artifacts complement the
// Inspector reads its "what did this Worker see, and what did it produce"
// answer from. Neither is a copy of the other — the audit response IS the
// server's replay.Timeline, byte-identical.

import type { Budget, Worker, Workflow } from './model'
import type { WFEvent } from './live'

export type NodeRecordState = 'pending' | 'running' | 'succeeded' | 'failed' | 'skipped'

/** NodeRecord mirrors core/replay.NodeRecord. Content is base64 (encoding/json's
 *  default for a Go []byte field) — decode with contentText/contentBytes below. */
export interface NodeRecord {
  state: NodeRecordState
  hash?: string
  type?: string
  content?: string
  costUsd?: number
  tokens?: number
  error?: string
}

/** Template mirrors core/server.Template — one row of GET /api/templates
 *  (M1.14's gallery): a `wee export` bundle's identity, nothing more (no
 *  UI-only/proprietary template format — the bundle IS the template). */
export interface Template {
  name: string
  workflowId: string
  version: string
  nodeCount: number
}

/** ImportedTemplate mirrors core/server.importTemplateResponse — POST
 *  /api/templates/{name}/import's response: where the unpacked files landed
 *  (for POST /api/run to resolve later) and the workflow itself, ready to
 *  hand straight to the workspace store's existing import path. */
export interface ImportedTemplate {
  workflowPath: string
  workflow: Workflow
}

/** ExecutionSummary mirrors core/server.ExecutionSummary — one row of GET
 *  /api/executions (M1.14's history table, REQ-METRIC-01/02). Cheap: no
 *  artifact bytes, just what the event log's ExecutionStarted/Finished and
 *  WorkerFinished events already carry. */
export interface ExecutionSummary {
  id: string
  workflow: string
  version: string
  state: string
  spentCostUsd: number
  spentTokens: number
  durationMs: number
}

/** Audit mirrors core/server.Audit (replay.Timeline plus the derived State). */
export interface Audit {
  executionId: string
  workflow: Workflow
  budget: Budget
  events: WFEvent[]
  nodes: Record<string, NodeRecord>
  spentCostUsd: number
  spentTokens: number
  definitionHashes?: Record<string, string>
  workers?: Record<string, Worker>
  state: string
}

/** contentText decodes a NodeRecord's base64 content as UTF-8 text — every
 *  artifact type the Inspector renders as text (Diff, Markdown, JSON, Code,
 *  TestResult, Report) goes through this. Returns undefined if there's no
 *  content yet (node pending/skipped) or it fails to decode. */
export function contentText(rec: NodeRecord | undefined): string | undefined {
  if (!rec?.content) return undefined
  try {
    return decodeURIComponent(escape(atob(rec.content)))
  } catch {
    return undefined
  }
}

/** contentDataURL builds a downloadable data: URL for a NodeRecord's raw
 *  bytes — the File artifact viewer's download link, and usable for any other
 *  type too (an image, an arbitrary Report). mimeType defaults to a generic
 *  binary type since domain.Artifact.mimeType isn't recorded on the wire yet
 *  (see core/domain/artifact.go — never actually populated at runtime). */
export function contentDataURL(rec: NodeRecord | undefined, mimeType = 'application/octet-stream'): string | undefined {
  if (!rec?.content) return undefined
  return `data:${mimeType};base64,${rec.content}`
}

/** contextHashesFor returns the artifact hashes a node's WorkerFinished event
 *  admitted as its resolved context (policy.Hashes, REQ-CTXPOL-03) — literally
 *  what the Worker saw, not a description of the policy. Empty if the node
 *  hasn't finished yet, was cached (no contextHashes recorded on that path —
 *  core/engine/node.go:236), or saw nothing (mode: none). */
export function contextHashesFor(audit: Audit, nodeId: string): string[] {
  const ev = audit.events.find((e) => e.type === 'WorkerFinished' && e.nodeId === nodeId)
  const hashes = ev?.payload?.contextHashes
  return Array.isArray(hashes) ? hashes.filter((h): h is string => typeof h === 'string') : []
}

/** nodeIdForHash finds which node in this execution produced a given artifact
 *  hash — how the Inspector resolves a context hash back to actual content:
 *  every hash a context policy can admit is some upstream node's own output,
 *  produced within the same execution and already in `audit.nodes`. */
export function nodeIdForHash(audit: Audit, hash: string): string | undefined {
  return Object.entries(audit.nodes).find(([, rec]) => rec.hash === hash)?.[0]
}

/** workerRef returns the "id@version" a node resolves to, or undefined for a
 *  tool-backed (or malformed) node — the key into Audit.workers. */
export function workerRef(audit: Audit, nodeId: string): string | undefined {
  return audit.workflow.nodes.find((n) => n.id === nodeId)?.worker
}

File: core/replay/audit.go
// Package replay renders and re-runs past executions (REQ-REPLAY-01..03). It
// is a thin layer over core/eventlog, core/store, and core/engine: Audit reads
// what an execution already wrote, at zero cost and zero model/tool calls;
// Reexecute runs the frozen snapshot again through the same Scheduler so the
// node cache (M1.6) naturally reuses unchanged nodes; Divergence compares two
// Timelines node by node. See docs/replay-honesty.md for what these two modes
// do and do not guarantee.
package replay

import (
	"fmt"

	"github.com/tzpereira/workflow-execution-engine/core/domain"
	"github.com/tzpereira/workflow-execution-engine/core/engine"
	"github.com/tzpereira/workflow-execution-engine/core/eventlog"
	"github.com/tzpereira/workflow-execution-engine/core/store"
)

// NodeRecord is one node's outcome as reconstructed purely from the event
// log — the same facts a live engine.Result reports (state, artifact
// identity, cost), plus the artifact's actual bytes, so the record is
// self-sufficient. State is one of engine's NodeState constants; a node with
// no recorded event is StateSkipped if the execution reached
// ExecutionFinished (it was never on the path taken), or StatePending
// otherwise (the run stopped before reaching it — crash or still in flight).
type NodeRecord struct {
	State   engine.NodeState    `json:"state"`
	Hash    string              `json:"hash,omitempty"`
	Type    domain.ArtifactType `json:"type,omitempty"`
	Content []byte              `json:"content,omitempty"` // base64 on the wire (encoding/json default for []byte)
	CostUSD float64             `json:"costUsd,omitempty"`
	Tokens  int64               `json:"tokens,omitempty"`
	Err     string              `json:"error,omitempty"`
}

// Timeline is a fully reconstructed execution: the frozen workflow it ran,
// every event in write order, and each node's final outcome. Nothing in a
// Timeline is re-run — it is exactly what Audit read from disk.
type Timeline struct {
	ExecutionID  string                `json:"executionId"`
	Workflow     domain.Workflow       `json:"workflow"`
	Budget       domain.Budget         `json:"budget"`
	Events       []domain.Event        `json:"events"`
	Nodes        map[string]NodeRecord `json:"nodes"`
	SpentCostUSD float64               `json:"spentCostUsd"`
	SpentTokens  int64                 `json:"spentTokens"`
	// DefinitionHashes are the content hashes of the definitions this execution
	// pinned at start (worker "id@version" → hash), read straight from the frozen
	// snapshot (REQ-VERSION-02). Audit never consults a registry — this is the
	// pinned record, immune to any registry change made since the run. nil if the
	// run pinned nothing (not registry-driven).
	DefinitionHashes map[string]string `json:"definitionHashes,omitempty"`
	// Workers is the full resolved Worker definition (goal, contract,
	// contextPolicy) behind each DefinitionHashes entry, pinned the same way
	// (REQ-UI-03) — the Inspector's source for a node's Contract.
	Workers map[string]domain.Worker `json:"workers,omitempty"`
	// Inputs is the resolved value (supplied or Default) behind every declared
	// Workflow.Inputs entry this execution actually used (REQ-INPUT-01) — what
	// this run was run against. Not a secret, so unlike a "${env:...}" value it
	// belongs here in the audit record.
	Inputs map[string]string `json:"inputs,omitempty"`
}

// Auditor renders past executions from their on-disk record alone (REQ-
// REPLAY-01): zero model calls, zero cost. It never constructs or touches a
// Scheduler or a NodeExecutor — Audit cannot reach a model or a tool even by
// accident, because it holds no reference to either.
type Auditor struct {
	log   *eventlog.Log
	store *store.Store
}

// NewAuditor builds an Auditor over the same event log and artifact store an
// engine.Scheduler wrote its executions to.
func NewAuditor(log *eventlog.Log, st *store.Store) *Auditor {
	return &Auditor{log: log, store: st}
}

// Audit reconstructs executionID's Timeline from its snapshot.json and
// events.jsonl (and the artifact store for each node's recorded content) —
// no network, no model, no tool invocation.
func (a *Auditor) Audit(executionID string) (Timeline, error) {
	var snap engine.Snapshot
	if err := a.log.ReadSnapshot(executionID, &snap); err != nil {
		return Timeline{}, fmt.Errorf("replay: audit %s: %w", executionID, err)
	}
	events, err := a.log.ReadAll(executionID)
	if err != nil {
		return Timeline{}, fmt.Errorf("replay: audit %s: %w", executionID, err)
	}

	nodes := make(map[string]NodeRecord, len(snap.Workflow.Nodes))
	finished := false
	for _, ev := range events {
		if ev.Type == domain.ExecutionFinished {
			finished = true
		}
		if ev.NodeID == "" {
			continue
		}
		rec := nodes[ev.NodeID]
		switch ev.Type {
		case domain.ArtifactCreated:
			if h, ok := ev.Payload["hash"].(string); ok {
				rec.Hash = h
			}
			if t, ok := ev.Payload["type"].(string); ok {
				rec.Type = domain.ArtifactType(t)
			}
		case domain.WorkerFinished:
			rec.State = engine.StateSucceeded
			if c, ok := ev.Payload["costUsd"].(float64); ok {
				rec.CostUSD = c
			}
			if t, ok := ev.Payload["tokens"].(float64); ok {
				rec.Tokens = int64(t)
			}
		case domain.Failure:
			rec.State = engine.StateFailed
			if e, ok := ev.Payload["error"].(string); ok {
				rec.Err = e
			}
		}
		nodes[ev.NodeID] = rec
	}

	var totalCost float64
	var totalTokens int64
	for _, n := range snap.Workflow.Nodes {
		rec, ok := nodes[n.ID]
		if !ok {
			state := engine.StatePending
			if finished {
				state = engine.StateSkipped
			}
			nodes[n.ID] = NodeRecord{State: state}
			continue
		}
		if rec.Hash != "" {
			content, err := a.store.Get(rec.Hash)
			if err != nil {
				return Timeline{}, fmt.Errorf("replay: audit %s: load artifact for node %q: %w", executionID, n.ID, err)
			}
			rec.Content = content
			nodes[n.ID] = rec
		}
		totalCost += rec.CostUSD
		totalTokens += rec.Tokens
	}

	return Timeline{
		ExecutionID:      executionID,
		Workflow:         snap.Workflow,
		Budget:           snap.Budget,
		Events:           events,
		Nodes:            nodes,
		SpentCostUSD:     totalCost,
		SpentTokens:      totalTokens,
		DefinitionHashes: snap.DefinitionHashes,
		Workers:          snap.Workers,
		Inputs:           snap.Inputs,
	}, nil
}
